README.md
"# DodecaDragons" 


changelog.txt
KNOWN BUGS:
-Font is often unreadable if it's centred due to weird antialiasing (deemed unsolvable, likely an browser-handled thing instead of a CSS thing)
-Font is also hard to read if the font size isn't a multiple of 12
-You can't save or export if you have an unusual character in your dragon's name
-Sigil automation tab apparently doesn't initially show until refresh for some people
-Plutonium tab disappears for some people
-Sometimes knowledge achievements aren't unlocked until the page is refreshed
-Knowledge reward range is sometimes slightly off for small amounts

TO DO:
-More hotkeys
-Change sigil upgrade max alls so that they use all of the resources
-Finally rearrange the stuff in the resource tab (or make it easier to do so)

v1.0.0 (09/07/23)
-The end of the game.
-Added finality cubes!
-Added a new dragon stage
-Changed the nuclear pasta info window to disappear immediately when mouse is moved
-Fixed finality essence upgrade 3 cost starting at 1.00e16 when it should start at 1.00e15
-Fixed knowledge upgrade 2 effect flickering when bought manually
-Fixed screen jolting around when tab key is pressed (tab indexing)
-Other minor bug fixes and changes

v0.12.2 (07/07/23)
-Added a new dragon stage
-Added some new finality essence upgrades and a new death essence upgrade
-Fixed knowledge sometimes getting stuck on 0
-Actually fixed formatting for time spent with dragon (as well as other cooldowns)

v0.12.1 (06/07/23)
-Added finality essence!
-Added a new nuclear pasta upgrade
-Fixed blue fire and holy fire upgrades flickering to 0 when being bought manually
-Fixed visual issue with one of the hypergod achievements
-Fixed death essence max all not spending the right amount
-Fixed console always displaying a source map error
-Somehow nobody caught that magic upgrade 5 didn't do what it said it did for like 8 months
-Other minor bug fixes and changes
-Removed Herobrine

v0.12.0 (05/07/23)
-Added nuclear pasta!
-Added max all and auto max all for cosmic plague upgrades (unlocks through achievements)
-Added max all for essence upgrades (unlocks through achievements)
-Fixed formatting for time spent with dragon
-Fixed errors with importing saves
-Other minor bug fixes and changes

v0.11.1 (02/07/23)
-Added death essence!
-Added a new dragon stage
-Fixed planet achievement 5 saying planet achievement 6's reward
-Fixed dark essence effect displaying incorrectly
-Fixed being able to input knowledge trade levels outside of the intended range
-Other minor bug fixes and changes

v0.11.0 (30/06/23)
-Added light and dark essences!
-Added a new oganesson upgrade
-Added Patreon link in info tab
-Improved the load function to allow for adding onto saved arrays when loading

v0.10.1 (28/06/23)
-Added oganesson!
-Added a new dragon stage
-Added a new cosmic plague upgrade
-Added buy max buttons for plague upgrades (unlocks through achievements)
-Dragon image can no longer be dragged
-Bought void magic upgrades are now less garish
-Fixed void magic unlock not checking for holy fire
-Other minor bug fixes and changes

v0.10.0 (26/06/23)
-Added the OMNIVERSAL HYPERGODS
-Added cosmic plague and spores
-Added buy max for planet forming (unlocks through achievements)
-Variables now update twice per second instead of once
-Fixed "next at" for dodecahedrons not disappearing when it should
-Other minor bug fixes and changes

v0.9.2 (22/06/23)
-Added superclusters!
-Added some more void magic upgrades
-Removed some old extraneous code
-Other minor bug fixes and changes

v0.9.1 (22/06/23)
-Added planets!
-Added some more void magic upgrades
-Added a new holy fire upgrade

v0.9.0 (21/06/23)
-Added holy dodecahedrons!
-Added void magic upgrades!
-Added some more holy octahedron upgrades
-Added a new dragon stage
-Removed Heresy as a hell layer option (it won't be added to the gameplay)
-Fixed holy fire upgrades saying "blue fire"
-Fixed blue fire upgrades being bought before blue fire is gained each tick
-Fixed sigil powers sometimes flickering to 0
-Hopefully fixed too many octahedron upgrades being purchasable when the tab first appears
-Other minor bug fixes and changes

v0.8.1 (12/02/23)
-Added holy octahedrons!
-Added holy fire!
-Added buy max button for tome upgrades (unlocks through achievements)
-Added auto max all for sigil upgrades (unlocks through achievements)
-Added a new hell layer!
-Removed achievement stars (they're extremely buggy)
-Clicking on a resource in the resource tab now takes you to that resource
-Fixed holy tetrahedron amount not updating in the resources tab
-Fixed some visual bugs in the holy tetrahedron tab
-Fixed knowledge trades not resetting on holy polyhedron reset
-Fixed sigil power sometimes going below 0
-Added a secret dragon Easter egg
-Other minor bug fixes and changes

v0.8.0 (01/02/23)
-Added holy tetrahedrons!
-Added achievement stars! Fully completed achievement rows now have some extra goals to reach
-Added hotkeys for activating/deactivating magic challenges and converting tomes
-Free uranium/plutonium gain is now based on highest ever to get, with platinum now also receiving it
-Some dark magic upgrade costs are cheaper
-Top left tabs can now be closed
-"Pet your dragon" button will now change to indicate when you cannot pet your dragon any more
-Some challenge-related QoL is attainable earlier
-Fixed blood gain exploding due to Greed not affecting gold/click
-Fixed orange sigil upgrade 4 not becoming inactive on reload if bought
-Fixed buying miners sometimes not working
-Fixed being able to use certain hotkeys before you unlock them
-Fixed platinum upgrade 7 becoming 20/10 in certain cases
-Other minor bug fixes and changes
-The game now also has some reworked sprites thanks to the talented Zaxolotl!

v0.7.4 (15/12/22)
-Added yellow sigils!
-Added max all and auto max all for blue fire upgrades (unlocks through achievements)
-Added auto max all for knowledge upgrades (unlocks through achievements)
-Added automatic plutonium gain (unlocks through achievements)

dodecaDragons/magic.js
//Unlocks magic
function unlockMagic() {
  if (game.gold.gte(1e15)) {
    game.gold = game.gold.sub(1e15)
    document.getElementById("unlockMagicButton").style.display = "none"
    document.getElementById("confirmationToggleBox").style.display = "block"
    document.getElementsByClassName("box")[6].style.display = "block"
    document.getElementsByClassName("box")[7].style.display = "block"
    document.getElementsByClassName("resourceRow")[4].style.display = "block"
    addUnlock() //sets unlock to 4
  }
}

//Confirmation popups for magic resets
function magicCheck() {
  if (game.magicToGet.eq(0)) {
    if (confirm("Woah, hold on! You won't gain any magic! Are you sure you want to perform a magic reset?")) magicReset()
  }
  else if (game.confirmations[0]) {
    if (confirm("Are you sure you want to perform a magic reset?")) magicReset()
  }
  else {magicReset()}
}

function magicReset(triggerLayer = "magic") {
  //Sets most pre-magic veriables back to their original states
  game.gold = new Decimal(0)
  if (game.unlockedAchievements[0] > 8) {game.miners = new Decimal(1)}
  else {game.miners = new Decimal(0)}
  game.minerCost = new Decimal(20)
  document.getElementById("minerCost").innerHTML = "20"
  
  game.fire = new Decimal(0)
  game.fireUpgrade1Bought = new Decimal(0)
  game.fireUpgrade1Cost = new Decimal(50)
  game.fireUpgrade2Bought = new Decimal(0)
  game.fireUpgrade2Cost = new Decimal(100)
  game.fireUpgrade3Bought = new Decimal(0)
  game.fireUpgrade3Cost = new Decimal(100)
  game.fireUpgrade4Bought = new Decimal(0)
  game.fireUpgrade4Cost = new Decimal(500)
  game.fireUpgrade5Bought = new Decimal(0)
  game.fireUpgrade5Cost = new Decimal(500)
  game.fireUpgrade6Bought = new Decimal(0)
  game.fireUpgrade6Cost = new Decimal(2e7)
  document.getElementById("fireUpgrade1Cost").innerHTML = format(game.fireUpgrade1Cost, 0)
  document.getElementById("fireUpgrade1Effect").innerHTML = format(new Decimal(2).pow(game.fireUpgrade1Bought.pow(0.6)), 2)
  document.getElementById("fireUpgrade2Cost").innerHTML = format(game.fireUpgrade2Cost, 0)
  document.getElementById("fireUpgrade3Cost").innerHTML = format(game.fireUpgrade3Cost, 0)
  document.getElementById("fireUpgrade4Cost").innerHTML = format(game.fireUpgrade4Cost, 0)
  document.getElementById("fireUpgrade5Cost").innerHTML = format(game.fireUpgrade5Cost, 0)
  document.getElementById("fireUpgrade6Cost").innerHTML = format(game.fireUpgrade6Cost, 0)
  document.getElementById("fireUpgrade6Effect").innerHTML = format(new Decimal(3).pow(game.fireUpgrade6Bought.pow(0.6)), 2)

  game.platinum = new Decimal(0)
  game.bestPlatinumToGet = new Decimal(0)
  if ( (triggerLayer === "magic" && game.unlockedAchievements[3] < 5) || (triggerLayer === "sigil" && game.unlockedAchievements[7] < 3) || (triggerLayer == "holyPolyhedron" && game.unlockedAchievements[20] < 2)) {
    for (i=0;i<9;i++) {
      if (i!=2 && i!=6) {
        game.platinumUpgradesBought[i] = 0
        document.getElementsByClassName("platinumUpgradesBought")[i].innerHTML = "0"
        document.getElementsByClassName("platinumUpgrade")[i].disabled = false
      }
    }
    document.getElementsByClassName("platinumUpgradesBought")[2].innerHTML = "1"
  }

  game.magic = game.magic.add(game.magicToGet)
  magicHardcap = new Decimal("e5000000")
  if (game.tomeUpgradesBought[8] == true) magicHardcap = magicHardcap.pow(game.blueFireUpgradesBought[3].pow(0.7).mul(5).add(1))
  if (game.magic.gt(magicHardcap)) {
    game.magic = magicHardcap
    document.getElementById("magicCap").innerHTML = " (hardcapped)"
  }
  else {document.getElementById("magicCap").innerHTML = ""}
}

//Magic upgrades
function buyMagicUpgrade(x) {
  //Checks that the magic upgrade is not bought and that the player's magic is greater/equal to the corresponding upgrade cost
  if (game.magicUpgradesBought[x-1] != true && game.magic.gte(magicUpgradeCosts[x-1])) {
    game.magic = game.magic.sub(magicUpgradeCosts[x-1])
    game.magicUpgradesBought[x-1] = true
    document.getElementsByClassName("magicUpgrade")[x-1].disabled = true
    if (x==8) document.getElementsByClassName("platinumUpgrade")[6].style.display = "block"
    if (x==11 && game.platinumUpgradesBought[6] == 10) document.getElementsByClassName("platinumUpgrade")[6].disabled = false 
    if (x==12) {
      document.getElementsByClassName("box")[8].style.display = "block"
      document.getElementsByClassName("resourceRow")[5].style.display = "block"
      document.getElementById("moreMagicUpgradesButton").style.display = "block"
      addUnlock() //sets unlock to 5
    }
    if (x==14) document.getElementById("fireUpgrade1Effect").innerHTML = format(new Decimal(3.5).pow(game.fireUpgrade1Bought.pow(0.7)), 2)
    if (x==19) document.getElementById("magifoldEffect").innerHTML = format(game.magifolds.pow(6), 2)
    if (x==20) {
      document.getElementsByClassName("magicUpgrade")[19].disabled = true
      document.getElementsByClassName("box")[9].style.display = "block"
      document.getElementsByClassName("resourceRow")[6].style.display = "inline-block"
      document.getElementById("morePUupgradesButton").style.display = "block"
      addUnlock() //sets unlock to 7
    }
  }
}

function magicUpgradeBuyMax() {for (i=1;i<21;i++) buyMagicUpgrade(i)}

//Unlocks more magic upgrades lol
function unlockMoreMagicUpgrades() {
  if (game.magifolds.gte(150000) && game.magic.gte(2e8)) {
    game.magic = game.magic.sub(2e8)
    document.getElementById("moreMagicUpgradesButton").style.display = "none"
    document.getElementsByClassName("box")[7].style.height = "560px"
    document.getElementsByClassName("box")[7].style.top = (window.innerHeight / 2 + renderVars.posY + renderVars.diffY - 133) + "px"
    for (i=12;i<20;i++) {document.getElementsByClassName("magicUpgrade")[i].style.display = "block"}
    addUnlock() //sets unlock to 6
  }
}

function unlockDarkMagicUpgrades() {
  if (game.gold.gte("e1000")) {
    game.gold = game.gold.sub("e1000")
    document.getElementById("unlockDarkMagicUpgradesButton").style.display = "none"
    document.getElementsByClassName("box")[10].style.display = "block"
    addUnlock() //sets unlock to 9
  }
}

//Dark magic upgrades
function buyDarkMagicUpgrade(x) {
  //Checks that the magic upgrade is not bought and that the player's magic is greater/equal to the corresponding upgrade cost
  if (game.darkMagicUpgradesBought[x-1] != true && game.magic.gte(darkMagicUpgradeCosts[x-1])) {
    game.magic = game.magic.sub(darkMagicUpgradeCosts[x-1])
    game.darkMagicUpgradesBought[x-1] = true
    document.getElementsByClassName("darkMagicUpgrade")[x-1].disabled = true
    if (x==4) document.getElementById("magifoldEffect").innerHTML = format(game.magifolds.pow(8), 2)
    if (x==8) {
      document.getElementById("horrorTabButton").style.display = "block"
      document.getElementsByClassName("box")[11].style.display = "block"
      document.getElementsByClassName("resourceRow")[7].style.display = "block"
      document.getElementsByClassName("confirmationToggle")[1].style.display = "inline-block"
      document.getElementById("cyanSigilUpgrade1Cost").innerHTML = format(game.cyanSigilUpgrade1Cost, 0)
      document.getElementById("cyanSigilUpgrade1Effect").innerHTML = format(game.cyanSigilUpgradesBought[0].add(1), 2)
      document.getElementById("cyanSigilUpgrade2Cost").innerHTML = format(game.cyanSigilUpgrade2Cost, 0)
      document.getElementById("cyanSigilUpgrade2Effect").innerHTML = format(game.cyanSigilUpgradesBought[1].add(1).pow(1.5), 2)
      document.getElementById("cyanSigilUpgrade3Cost").innerHTML = format(game.cyanSigilUpgrade3Cost, 0)
      document.getElementById("cyanSigilUpgrade3Effect").innerHTML = format(new Decimal(1.2).pow(game.cyanSigilUpgradesBought[2].pow(0.4)), 2)
      if (game.cyanSigilUpgradesBought[3] == 1) {document.getElementsByClassName("cyanSigilUpgrade")[3].disabled = true}
      else {document.getElementsByClassName("cyanSigilUpgrade")[3].disabled = false}
      addUnlock() //sets unlock to 10
    }
  } 
}

function darkMagicUpgradeBuyMax() {
  for (i=1;i<9;i++) buyDarkMagicUpgrade(i)
  if (game.pinkSigilUpgradesBought[1].eq(1)) {for (i=9;i<21;i++) buyDarkMagicUpgrade(i)}
}

function moreDarkMagicUpgrades() {
  if (game.gold.gte("e6.66e16")) {
    game.gold = game.gold.sub("e6.66e16")
    document.getElementById("moreDarkMagicUpgradesButton").style.display = "none"
    for (i=10;i<20;i++) document.getElementsByClassName("darkMagicUpgrade")[i].style.display = "inline-block"
    addUnlock() //sets unlock to 19
  }
}

function unlockVoidMagicUpgrades() {
  if (game.gold.gte("e1e100") && game.holyFire.gte(1e8)) {
    document.getElementById("unlockVoidMagicUpgradesButton").style.display = "none"
    document.getElementsByClassName("box")[34].style.display = "block"
    document.getElementsByClassName("upgradeDragonButton")[6].style.display = "block"
    addUnlock() //sets unlock to 27
  }
}

//Void magic upgrades
function buyVoidMagicUpgrade(x) {
  //Checks that the magic upgrade is not bought and that the player's magic is greater/equal to the corresponding upgrade cost
  if (game.voidMagicUpgradesBought[x-1] != true && game.magic.gte(voidMagicUpgradeCosts[x-1])) {
    game.voidMagicUpgradesBought[x-1] = true
    document.getElementsByClassName("voidMagicUpgrade")[x-1].disabled = true
    if (x==4) {
      document.getElementsByClassName("box")[35].style.display = "block"
      document.getElementsByClassName("box")[36].style.display = "block"
      document.getElementsByClassName("resourceRow")[23].style.display = "block"
      holyDodecahedronUnlockCheck()
      addUnlock() //sets unlock to 28
    }
		else if (x==8) {
			document.getElementsByClassName("holyFireUpgrade")[5].style.display = "inline-block"
			document.getElementsByClassName("holyFireBuyMaxButton")[5].style.display = "inline-block"
			game.holyFireUpgradeCosts[5] = new Decimal(holyFireUpgradeBase[6]).pow(game.holyFireUpgradesBought[5]).mul(holyFireUpgradeInitialCosts[6]).floor()
			document.getElementById('holyFireUpgrade6Cost').innerHTML = format(game.holyFireUpgradeCosts[5], 0);
		}
		else if (x==16) {
      document.getElementsByClassName("box")[39].style.display = "block"
			document.getElementsByClassName("resourceRow")[25].style.display = "block"
			document.getElementsByClassName("upgradeDragonButton")[7].style.display = "block"
      addUnlock() //sets unlock to 31
    }
		else if (x==17) {
      document.getElementsByClassName("box")[40].style.display = "block"
			document.getElementsByClassName("resourceRow")[26].style.display = "block"
			document.getElementById("unlockEssencesButton").style.display = "block"
      addUnlock() //sets unlock to 32
    }
  } 
}

function getMagicGain() {
  let toGet = game.gold.div(1e15).pow(0.1);
  if (game.magicUpgradesBought[3]) {
    if (game.magicUpgradesBought[9]) {toGet = toGet.mul(game.magic.add(1).log2().mul(4).add(1))}
    else {toGet = toGet.mul(game.magic.add(1).log2().mul(1.5).add(1))}
  }
  if (game.magicUpgradesBought[1]) toGet = toGet.mul(1.5);
  if (game.magicUpgradesBought[6]) toGet = toGet.mul(2);
  if (game.magicUpgradesBought[17]) toGet = toGet.mul(3);
  if (game.unlocks >= 10) toGet = toGet.mul(game.cyanSigils.add(1).pow(2));
  if (game.unlocks >= 11) toGet = toGet.mul(game.blueSigils.add(1).pow(3));
  if (game.unlocks >= 12) toGet = toGet.mul(game.indigoSigils.add(1).pow(4));
  if (game.unlocks >= 13) toGet = toGet.mul(game.violetSigils.add(1).pow(5));
  if (game.unlocks >= 14) toGet = toGet.mul(game.pinkSigils.add(1).pow(6));
  if (game.darkMagicUpgradesBought[9] == true) toGet = toGet.pow(1.1);
  if (game.darkMagicUpgradesBought[1]) toGet = toGet.pow(game.uranium.add(1).log10().div(30).add(1));
  magicHardcap = new Decimal("e5000000")
  if (game.tomeUpgradesBought[8] == true) magicHardcap = magicHardcap.pow(game.blueFireUpgradesBought[3].pow(0.7).mul(5).add(1))
  if (toGet.gt(magicHardcap)) toGet = magicHardcap
  toGet = toGet.floor();
  return toGet;
}

function getNextMagicAt() {
  let nextMagic = game.magicToGet.add(1);
  if (game.magicUpgradesBought[17]) nextMagic = nextMagic.div(3);
  if (game.magicUpgradesBought[6]) nextMagic = nextMagic.div(2);
  if (game.magicUpgradesBought[1]) nextMagic = nextMagic.div(1.5);
  if (game.magicUpgradesBought[3]) {
    if (game.magicUpgradesBought[9]) {nextMagic = nextMagic.div(game.magic.add(1).log2().mul(4).add(1))}
    else {nextMagic = nextMagic.div(game.magic.add(1).log2().mul(1.5).add(1))}
  }
  nextMagic = nextMagic.pow(10).mul(1e15).add(1);
  if (game.unlocks >= 10) nextMagic = new Decimal(6969); //should be hidden by this point, so nobody should see this
  return nextMagic;
}